source/common/router/router.cc

```diff
@@ -844,11 +844,33 @@ void Filter::cleanup() {
   }
 }
 
+std::string Filter::getShadowCluster(const ShadowPolicy& policy,
+                                     const Http::HeaderMap& headers) const {
+  if (!policy.cluster().empty()) {
+    return policy.cluster();
+  } else {
+    ASSERT(!policy.clusterHeader().get().empty());
+    const auto entry = headers.get(policy.clusterHeader());
+    if (!entry.empty()) {
+      return std::string(entry[0]->value().getStringView());
+    }
+    ENVOY_STREAM_LOG(warn, "There is no cluster name in header: {}", *callbacks_,
+                     policy.clusterHeader());
+    return "";
+  }
+}
+
 void Filter::maybeDoShadowing() {
   for (const auto& shadow_policy_wrapper : active_shadow_policies_) {
     const auto& shadow_policy = shadow_policy_wrapper.get();
 
-    ASSERT(!shadow_policy.cluster().empty());
+    const std::string cluster = getShadowCluster(shadow_policy, *downstream_headers_);
```

Comment:
What ensures that the cluster name stated by the header exists?
Should we skip the rest of this iteration if the cluster is not found?

Reply:
Added the empty string check in `getShadowCluster` for this scenario `some_cluster_header: ""` in headers. If the cluster name got from headers is empty, we skip the rest of this iteration. Other than that, there is an additional check to see if the cluster is in cluster manager in the `shadow` [method](https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/source/common/router/shadow_writer_impl.cc#L19-L23).